openml/runs/run.py

```diff
@@ -156,58 +152,86 @@ def _create_description_xml(self):
                                detailed_evaluations=self.detailed_evaluations,
                                tags=self.tags)
         description_xml = xmltodict.unparse(description, pretty=True)
         return description_xml
 
     @staticmethod
-    def _parse_parameters(model, server_flow):
-        """Extracts all parameter settings from a model in OpenML format.
+    def _parse_parameters(flow, model=None):
+        """Extracts all parameter settings from the model inside a flow in
+        OpenML format.
 
         Parameters
         ----------
-        model
-            the scikit-learn model (fitted)
-        flow
+        flow : OpenMLFlow
             openml flow object (containing flow ids, i.e., it has to be downloaded from the server)
 
+        model : BaseEstimator, optional
+            If not given, the parameters are extracted from ``flow.model``.
+
         """
-        if server_flow.flow_id is None:
-            raise ValueError("The flow parameter needs to be downloaded from server")
+
+        if model is None:
+            model = flow.model
+
+        openml.flows.functions._check_flow_for_server_id(flow)
 
         def get_flow_dict(_flow):
             flow_map = {_flow.name: _flow.flow_id}
             for subflow in _flow.components:
                 flow_map.update(get_flow_dict(_flow.components[subflow]))
             return flow_map
 
-        def extract_parameters(_flow, _param_dict, _main_call=False, main_id=None):
+        def extract_parameters(_flow, _flow_dict, component_model,
+                               _main_call=False, main_id=None):
             # _flow is openml flow object, _param dict maps from flow name to flow id
             # for the main call, the param dict can be overridden (useful for unit tests / sentinels)
-            # this way, for flows without subflows we do not have to rely on _param_dict
+            # this way, for flows without subflows we do not have to rely on _flow_dict
             _params = []
             for _param_name in _flow.parameters:
                 _current = OrderedDict()
                 _current['oml:name'] = _param_name
-                _current['oml:value'] = _flow.parameters[_param_name]
+
+                _tmp = openml.flows.sklearn_to_flow(
+                    component_model.get_params()[_param_name])
+
+                # Try to filter out components which are handled further down!
+                if isinstance(_tmp, openml.flows.OpenMLFlow):
+                    continue
+                try:
+                    _tmp = json.dumps(_tmp)
+                except TypeError as e:
+                    # Python3.5 exception message:
+                    # <openml.flows.flow.OpenMLFlow object at 0x7fed87978160> is not JSON serializable
+                    # Python3.6 exception message:
+                    # Object of type 'OpenMLFlow' is not JSON serializable
+                    if 'OpenMLFlow' in e.args[0] and \
```

Comment:
in the catch, please try to define what can reasonably fall into that (because we handle it further down) and raise exception if we have something unexpected